ml/jml/fasttext_generator.cc

```diff
@@ -0,0 +1,236 @@
+
+    //Build feature map
+    for (size_t f = 0; f < features.size(); ++f) {
+        output->featureMap[features[f]] = f;
+    }
+
+    //0 : Initialize
+    int64_t ntokens = 0;
+    for (int i = 0; i < training_data.example_count(); ++i) {
+        const Feature_Set & featureSet = training_data[i];
+
+        auto it = featureSet.begin();
+        auto itEnd = featureSet.end();
+
+        while (it != itEnd) {
+            std::pair<Feature, float> feature = *it;
+            if (feature.first == predicted) {
+                labelCount[feature.second]++;
+            }
+            else {
+                ntokens += (int64_t)feature.second;
+            }
+            ++it;
+        }
+    }
+
+    args_->model = fasttext::model_name::sup;
+    args_->thread = training_data.example_count() < numCpus() ? 1 : numCpus();
+    args_->bucket = 0;
+    args_->dim = dims;
+    args_->epoch = epoch;
+    args_->verbose = verbose;
+
+
+    std::shared_ptr<fasttext::Matrix> input_ = std::make_shared<fasttext::Matrix>(features.size()+args_->bucket, args_->dim);
+    input_->uniform(1.0 / args_->dim);
+    std::shared_ptr<fasttext::Matrix> output_ = std::make_shared<fasttext::Matrix>(numLabels, args_->dim);   
+    output_->zero();
+    fastTextModel.input_ = input_;
+    fastTextModel.output_ = output_;
+
+    //Note that we skip Fasttext's dictionary because the strings are already abstracted in our features  
+
+    std::atomic<int64_t> tokenCount(0);
+    auto trainThread = [&] (size_t threadId) {
+
+      int32_t start = threadId * training_data.example_count() / args_->thread;
+
+      fasttext::Model model(input_, output_, args_, threadId);
+      model.setTargetCounts(labelCount);
+    
+      int64_t localTokenCount = 0;
+      std::vector<int32_t> line, labels;
+      while (tokenCount < args_->epoch * ntokens) {
+
+        //if we reach the end we wrap around
+        if (start >= training_data.example_count())
+            start = 0;
+
+        fasttext::real progress = fasttext::real(tokenCount) / (args_->epoch * ntokens);
+        fasttext::real lr = args_->lr * (1.0 - progress);
+
+        const Feature_Set & lineFeatureSet = training_data[start];
+        auto it = lineFeatureSet.begin();
+        auto itEnd = lineFeatureSet.end();
+
+        line.clear();
+        labels.clear();
+
+        while (it != itEnd) {
+            std::pair<Feature, float> feature = *it;
+            if (feature.first == predicted) {
+                labels.push_back(feature.second); //value of the label feature is the label
+            }
+            else {
+                auto featureit = output->featureMap.find(feature.first);
+                if (featureit != output->featureMap.end()) {
+                    size_t f = (*featureit);
+                    for (int i = 0; i < feature.second; ++i)
+                        line.push_back(f);        
+                }
+            }
+            localTokenCount++;
+            ++it;
```

Comment:
it is an iterator so using prefix is slightly more efficient. localTokenCount is an atomic so it doesnt matter.